pkg/tls/service.go

```diff
@@ -53,6 +59,33 @@ func (s *Service) GetCertificateSecret(ctx context.Context, host string, namespa
 	return tlsSecret, nil
 }
 
+func (s *Service) CleanupCertificates(ctx context.Context, owner interface{}) error {
+	gateway, ok := owner.(*gatewayv1beta1.Gateway)
+	if ok {
+		// get names of hosts for traffic object being deleted
+		trafficAccessor := traffic.NewGateway(gateway)
+		for _, host := range trafficAccessor.GetHosts() {
+			secret := &v1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      host,
+					Namespace: gateway.Namespace,
+				},
+			}
+			err := s.controlClient.Delete(ctx, secret)
+			if err != nil && !k8serrors.IsNotFound(err) {
+				return fmt.Errorf("error deleting cert secret: %s", err)
+			}
+		}
+		return nil
+	}
+	// TODO ingress case
+	_, ok = owner.(*v12.Ingress)
+	if ok {
+		return nil
+	}
+	return fmt.Errorf("uknkown object type for a certificate deletion: %s", reflect.TypeOf(owner))
+}
```

Comment:
Same here, use  traffic.Interface instead of interface{}, and make the reconciler(gateway, ingress) pass it in.

Reply:
gateway type does not implement `traffic.interface` if I'm not mistaken